### DISCUSSION TITLE
Delete flow behavior

### DISCUSSION THREAD

**Discussion by roytman** [fybrik/fybrik#1471] [Developers]
When a Fybrik application creates a WRITE flow and doesn't specify an existing catalog asset with a destination bucket, it creates a bucket.
In order to support API symmetry, the DELETE flow should remove empty buckets created by Fybrik. 
The proposed implementation is:

- when Fybrik creates a new bucket, its catalog entry will be marked with a dedicated tag/flag, e.g. `fybrik-created`. Question: is catalog tag good enough, or we can/should mark the bucket/objects (e.g. `Custom metadata` )? 
- during the DELETE operation, when the data asset is removed, if the bucket is empty and has the mentioned above tag, Fybrik will remove the bucket too. Otherwise, the bucket will not be removed.

**Comment by shlomitk1** [comment]
Creation/deletion of a bucket is done using datashim. 
There are two options here:
1. We discuss with the team the new requirements, including the minimal deployment/functionality as we don't need PVCs that are provided as well.
2. We stop using datashim and implement bucket creation/deletion on our own.

**Reply by flora177** [reply]
Coping here Sima's answer to how to tag the fybrik created storage: 
It will likely be different for different storage technologies.  We could either label the storage somehow, or fybrik knows which storage it manages.  It has storage CRDs for this.

**Reply by Raviv-S** [reply]
@revit13 and I were wondering if bucket created with datashim won't be deleted if we won't use datashim to delete it.
What I mean is if there is a mechanism in datashim implementation that recreate the bucket if it was deleted from outside source.

**Reply by shlomitk1** [reply]
No, it does not recreate the bucket. Normally, deleting Dataset would delete the relevant bucket. We asked to add an option (some label in the resource) that won't cause datashim to delete the bucket upon deleting Dataset, that's all.

**Reply by Raviv-S** [reply]
The module at the moment deletes the bucket directly with s3 api (if the bucket is empty after the object deletion and a deleteEmptyBucket flag was set), without using datashim or addressing Datasets

**Comment by revit13** [comment]
The delete module that @Raviv-S is working on can do the actual deletion: if we have a policy defined as follows, such that the module could get the `DeleteEmptyBucketAction` action in its helm values if the asset has a tag `deleteEmptyBucket`. this tag can be set by the user or the manager if it created such bucket. I am not familiar with datashim so I am not sure what information the module needs to delete `Dataset resource` created by the manager. 

```
    rule[{"action": {"name":"DeleteEmptyBucketAction"}, "policy": description}] {
      description := "delete empty buckets"
      input.resource.metadata.tags.deleteEmptyBucket
      input.action.actionType == "delete"
    }

```

**Reply by shlomitk1** [reply]
The `Dataset` resource may no longer exist.

**Comment by roytman** [comment]
@shlomitk1  and I had a discussion about possible solutions, as result Shlomit will open 2 issues:

1.  replace datashim by our direct S3 API access: [1499](https://github.com/fybrik/fybrik/issues/1499)
2. add an option to configure delete behavior via policies. The delete module will get a single flag delete or not empty buckets: [1500](https://github.com/fybrik/fybrik/issues/1500)

**Comment by simanadler** [comment]
Some thoughts:
1) @shlomitk1 @roytman Our current implementation supports only S3 buckets.  We need to broaden this to support other types of storage, so I'm not sure that moving to S3 API access is the right approach.  There exist proprietary solutions for this.  Need to check if datashim or other open source has matured to support additional types of data stores.

2) Currently creation of the bucket is done by the manager.  Does it make sense to have creation done by the manager and delete by a module?  Seems we should have a more symmetrical approach.

3) @shlomitk1 Regarding adding delete behavior via policies, I'm not sure why since the create is not policy based.
